server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java

```diff
@@ -3781,6 +3781,51 @@ protected long doGenerateSeqNoForOperation(Operation operation) {
         }
     }
 
+    public void testSegmentContainsOnlyNoOps() throws Exception {
+        Engine.NoOpResult noOpResult = engine.noOp(new Engine.NoOp(1, primaryTerm.get(), REPLICA, randomNonNegativeLong(), "test"));
```

Comment:
primaries can also have noops upon promotion/recovery. Randomize the engine type?